File: W2D3/tic_tac_toe_project/tic_tac_toe_v1/board.rb
class Board

    def initialize
        @grid = Array.new(3) {Array.new(3, "_")}
    end

    def [](position)
        row, col = position
        @grid[row][col]
    end

    def []=(position, value)
        row, col = position
        @grid[row][col] = value
    end

    # def valid?(position)
    #     row, col = position
    #     position.all? do |i|
    #         i >= 0 && i < @grid.length
    #     end
    # end

    def valid?(position)
        row, col = position
        position.each do |i|
            if i < 0 || i > @grid.length
                return false
            end
        end
        true
    end

    def empty?(position)
        self[position] == "_"
    end

    def place_mark(position, mark)
        if !valid?(position) || !empty?(position)
            raise "invalid mark"
        end
        self[position] = mark
    end

    def print
        @grid.each do |row|
            puts row.join(" ")
        end
    end

    def win_row?(mark)
        @grid.any? do |row|
            row.all?(mark)
        end
    end

    def win_col?(mark)
        @grid.transpose.any? do |col|
            col.all?(mark)
        end
    end

    def win_diagonal?(mark)
        left = (0...@grid.length).all? do |i|
            position = [i, i]
            self[position] == mark
        end
        right = (0...@grid.length).all? do|i|
            row = i
            col = @grid.length - 1 - i
            position = [row, col]
            self[position] == mark
        end
        if left || right
            return true
        else
            return false
        end
    end

    def win?(mark)
        win_row?(mark) || win_col?(mark) || win_diagonal?(mark)
    end

    def empty_positions?
        @grid.each do |row|
            row.each do |ele|
                if ele == "_"
                    return true
                end
            end
        end
        false
    end

end

File: W8D4/JS_TTT/board.js
function Board() {
    this.grid = new Array(3).fill(0).map(() => new Array(3));
}

Board.prototype.won = function() {
    for(let i = 0; i < this.grid.length; i++) {
        if (this.grid[i][0] === this.grid[i][1] && this.grid[i][1] === this.grid[i][2]) {
            return true;
        } else if (this.grid[0][i] === this.grid[1][i] && this.grid[1][i] === this.grid[2][i]){
            return true;
        } 
    }
    if (this.grid[0][0] === this.grid[1][1] && this.grid[1][1] === this.grid[2][2]) {
        return true;
    }

    if (this.grid[0][2] === this.grid[1][1] && this.grid[1][1] === this.grid[2][0]) {
        return true;
    }

    return false;

}

Board.prototype.winner = function() {
    if (this.won) {
        return true;
    } else {
        return false;
    }
}

Board.prototype.validPos = function(pos) {
    let [row, col] = pos;
    if ((row > this.grid.length || row < 0) && (col > this.grid.length || col < 0)) {
        return false;
    } else {
        return true;
    }
}
Board.prototype.empty = function(pos) {
    let [row,col] = pos;
    if (this.grid[row][col] !== undefined) {
        return false;
    }
    return true;
}

Board.prototype.place_mark = function(pos, mark) {
    let [row, col] = pos;
    if (this.empty(pos) && this.validPos(pos)) {
        this.grid[row][col] = mark;
    } else {
        throw new Error('not a valid position');
    }
}